### ISSUE TITLE
"ERROR in Ng2PaginationModule is not an NgModule" after compiling succesfully

### ISSUE THREAD

**Issue by wsloth** [michaelbromley/ngx-pagination#114]
**Angular 2 version**: 2.3.1

**ng2-pagination version**: 2.0.0

**Description of issue**: 
After importing the Ng2PaginationModule in my `app.module.ts`, everything works, but after building (angular-cli@latest) it outputs an error: `ERROR in Ng2PaginationModule is not an NgModule`

**Steps to reproduce**:
Simply install using NPM and use in your application.
```
import { Ng2PaginationModule } from 'ng2-pagination';

...

@NgModule({
  imports: [
    BrowserModule,
    Ng2PaginationModule,
    FormsModule,
    HttpModule
  ],
...
```

**Expected result**:
Everything should compile without errors. Actually, it compiles and outputs all bundle files, but after it's done it still throws an error, messing with my teamcity build.

**Comment by michaelbromley** [comment]
So it actually works in the app, but just errors in the build? Strange. Is this happening just with AoT builds or JiT too?

**Comment by wsloth** [comment]
Yeah, it works perfectly (kudos to you, amazing little library and clever source!), it just throws an error after building, which is totally strange. This happens in AOT compilation (the angular-cli default).

It shows the build process, and then the outputted bundle files. After it's outputted the "succesfully completed" overview (of files generated) it throws the error.

**Comment by wsloth** [comment]
I have been digging around and it probably has something to do with the module being configurated wrongly:

This comment by hansl: https://github.com/angular/angular-cli/issues/3426#issuecomment-269673735
In which he links to [this article.](https://medium.com/@cyrilletuzi/how-to-build-and-publish-an-angular-module-7ad19c0b4464#.4y2l9v7bk)

Could it be because it's missing a rollup file [as seen in this template?](https://github.com/beeman/generator-ngx-module/blob/master/generators/app/templates/_rollup.config.js)

**Comment by michaelbromley** [comment]
Thanks - just had a glance over the issue and article. Looks useful - for a start the article says that modules should be published in UMD format, whereas I currently just publish CommonJS.

When I get a bit more time to work on this, I'll look into using rollup as in the example.

**Comment by wsloth** [comment]
If it's as simple as adding a rollup file it should be easy, I'll take a look at it as well when I have time for it. I've never worked on publishable angular modules though so I have no idea what I'm doing 😃

**Comment by michaelbromley** [comment]
> I have no idea what I'm doing

Haha, I know how that feels.

**Comment by andrejkaurin** [comment]
I have the same issue and ng build -prod fails because

**Comment by jonigl** [comment]
I have the same issue here using `ng build --aot --prod `

**Comment by michaelbromley** [comment]
Please see v3.0.0 (now published as ngx-paginate). It is bundled by Rollup in ES module and UMD formats.

**Comment by Pinank** [comment]
Same issue on 
ng build --prod

Any Suggestion for the solution?

**Comment by michaelbromley** [comment]
@Pinank I just tested this:

* installed Angular CLI and created a new project.
* in that project I installed `ngx-pagination` (v3.0.0)
* imported the `NgxPaginationModule` and used the pipe and component in the app component.
* ran both `ng build --prod` and `ng build --aot --prod` and both of them completed without error.
* checked in the browser and everything works.

If you can tell me how to reproduce the error that you see, I'll investigate.

**Comment by NYearOldCoder** [comment]
I've created a simple Angular project in VS Code with Angular CLI v1.0.0, ngx-pagination v3.0.0 and cannot build with `--prod` and/or `--aot` options.

- `ng build` no errors
- `ng build --prod` failed
- `ng build --aot` failed

You can download ZIP from DropBox (sorry, no Plunker):
[https://www.dropbox.com/s/a50ny382u152ctb/pagination-test.zip]

**Comment by michaelbromley** [comment]
@NYearOldCoder I just did the followingL

* downloaded and extracted your zip file
* ran `npm install`
* ran `ng build --prod`

And got the following error
```
ERROR in ng:///C:/Development/temp/pagination-test/src/app/list/list.component.html (3,36): Property 'p' does not exist on type 'ListComponent'.
ng:///C:/Development/temp/pagination-test/src/app/list/list.component.html (3,3): Property 'p' does not exist on type 'ListComponent'.
```
Is this the same error you get? The reason for this error is that in list.component.html template, you refer to a variable "p" which was not defined on the component class. As soon as I add a `p = 1;` property to the class, everything builds without error.

**Comment by NYearOldCoder** [comment]
That's right. I've got the same error. It's not clear to me why `ng build` command has succeeded.
Thanks.

**Comment by michaelbromley** [comment]
@NYearOldCoder The reason is that `ng build` does not do the step of compiling your templates into TypeScript, which does happen then you use the `--prod` flag. Once a template is compiled into a TypeScript class, it becomes apparent that the variable "p" does not exist on the component, leading to the error.

In dev mode (`ng build`) this compilation step happens dynamically at run-time, and therefore the TypeScript compiler does not get a chance to complain about it.

**Comment by NYearOldCoder** [comment]
Thank you for the explanation.
I think a lot of people would appreciate if you include that variable declaration and initialization in your Simple Example.

**Comment by michaelbromley** [comment]
@NYearOldCoder You are totally correct. I have done that now 👍

**Comment by beligh-hamdi** [comment]
<pagination-controls (pageChange)="p = $event"></pagination-controls>

export class XComponent  {
   ...
  p: number;

just add p in your typescript, add should be work. (test it on my app with ng cli 1.0.0à
ng build --prod

**Comment by zap-deepak** [comment]
ERROR in Ng2PaginationModule is not an NgModule
in angular cli.

**Comment by michaelbromley** [comment]
I am closing this now since I believe the issue to be fixed with version 3.0.0 and above. Please let me know if you have any further issues and I'll re-open it.

**Comment by signalhub** [comment]
ERROR in Error: NgxPaginationModule is not an NgModule
